### PR TITLE
Add `hab file` commands and encryption

### PR DESCRIPTION
**Pull request by adamhjk**
* Adds in the `hab file` commands
* Turns security and encryption on for both `config apply` and `file
  upload`
* Breaks the `config apply` and `file upload` commands into a separate
  `hab-butterfly` component and command.
* Fixes the order of arguments to be the same between file upload and
  config apply.
* Allows for encryption of configuration as well as files.
* Improves the `butterfly` http gateway output to render as strings if
  the contents are UTF-8.
* Allows for the hashing of a pure byte buffer.
* Failed initialization will now block starting/restarting.

gif-keyboard-15632042006024311536

Signed-off-by: Adam Jacob <adam@chef.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/butterfly/src/message/mod.rs [hunk 1]
```diff
@@ -138,7 +139,10 @@ impl Encodable for swim::ServiceConfig {
             try!(s.emit_struct_field("service_group", 0, |s| self.get_service_group().encode(s)));
             try!(s.emit_struct_field("incarnation", 1, |s| self.get_incarnation().encode(s)));
             try!(s.emit_struct_field("encrypted", 2, |s| self.get_encrypted().encode(s)));
-            try!(s.emit_struct_field("config", 3, |s| self.get_config().encode(s)));
+            match str::from_utf8(self.get_config()) {
```
Comment (reset) [inline components/butterfly/src/message/mod.rs]: Nice work here! Much better

File: components/butterfly/src/message/mod.rs [hunk 2]
```diff
@@ -152,7 +156,10 @@ impl Encodable for swim::ServiceFile {
             try!(s.emit_struct_field("incarnation", 1, |s| self.get_incarnation().encode(s)));
             try!(s.emit_struct_field("encrypted", 2, |s| self.get_encrypted().encode(s)));
             try!(s.emit_struct_field("filename", 3, |s| self.get_filename().encode(s)));
-            try!(s.emit_struct_field("body", 4, |s| self.get_body().encode(s)));
+            match str::from_utf8(self.get_body()) {
```
Comment (reset) [inline components/butterfly/src/message/mod.rs]: Same here, much better for the end user

**Comment by reset** [comment]
@thesentinels approve

**Comment by thesentinels** [comment]
:metal: I am testing your branch against master before merging it. We do this to ensure that the master branch is never failing tests.

**Comment by thesentinels** [comment]
:neckbeard: Travis CI has [started testing this PR](https://travis-ci.org/habitat-sh/habitat/builds/179512213).

**Comment by thesentinels** [comment]
:sparkling_heart: Travis CI [reports this PR passed](https://travis-ci.org/habitat-sh/habitat/builds/179512213).

It always makes me feel nice when humans approve of one anothers work. I'm merging this PR now.

I just want you and the contributor to answer me one question:

gif-keyboard-3280869874741411265

**Pull request merged**